File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/firebase/firestore/transactions/FirestoreUser.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.firebase.firestore.transactions;

import android.util.Log;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.HashMap;
import java.util.Map;

public class FirestoreUser {

    private FirebaseFirestore db = FirebaseFirestore.getInstance();

    private final String usersCollection = "Users";

    private final String accountCollection = "Account";
    private final String purchaseHistoryCollection = "PurchaseHistory";

    private final String creditsCollection = "Credits";
    private final String purchasesCollection = "PurchaseHistory";

    private final String fieldUID = "uid";

    public FirebaseUser getCurrentFirebaseUser() {
        return FirebaseAuth.getInstance().getCurrentUser();
    }

    public void createUserRecord() {
        /*
        db.collection("Users")
                .document(getCurrentFirebaseUser().getUid())
                .collection(accountCollection)
                .document(creditsCollection)
                .set(accountCredits, SetOptions.merge())
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        Log.d("Firestore", "DocumentSnapshot successfully written!");
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.w("Firestore", "Error writing document", e);
                    }
                });
        */

        db.collection("Users")
                .document(getCurrentFirebaseUser().getUid())
                .get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot documentSnapshot) {
                        if(documentSnapshot.exists()) {
                            Log.d("Firestore", "Document " + getCurrentFirebaseUser().getUid() + " successfully exists!");
                        } else {
                            Log.d("Firestore", "Document " + getCurrentFirebaseUser().getUid() + " successfully created!");
                            initUserDoc(documentSnapshot.getReference());
                        }
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.d("Firestore", "Document" + getCurrentFirebaseUser().getUid() + " already exists!");
                    }
                })
                .addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<DocumentSnapshot> task) {
                        Log.d("Firestore", "Process completed.");
                    }
                });
    }

    public void initUserDoc(DocumentReference doc) {
        Map<String, Object> accountCredits = new HashMap<>();
        Map<String, Object> purchaseHistory = new HashMap<>();

        Map<String, Object> data = new HashMap<>();
        data.put(creditsCollection, accountCredits);
        data.put(purchasesCollection, purchaseHistory);

        doc.collection(accountCollection)
            .document(creditsCollection)
                .getParent().document(purchaseHistoryCollection);
    }

    public DocumentReference getUserDoc() {
        if(getCurrentFirebaseUser() == null) {
            return null;
        }

        //return db.collection(userTable).document(getCurrentFirebaseUser().getUid());
        return db.collection(usersCollection).document(getCurrentFirebaseUser().getUid());
    }

    public CollectionReference getUserCollection() {
        if(getCurrentFirebaseUser() == null) {
            return null;
        }

        //return db.collection(userTable).document(getCurrentFirebaseUser().getUid());
        return db.collection(usersCollection);
    }

}

File: app-compose/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/data/firebase/firestore/transactions/user/FirestoreUser.kt
package com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user

import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.account.FirestoreAccount
import org.jetbrains.annotations.TestOnly

class FirestoreUser {
    companion object {
        private const val COLLECTION_USERS = "Users"

        val currentFirebaseUser: FirebaseUser?
            get() = FirebaseAuth.getInstance().currentUser

        val userDocument: DocumentReference
            @Throws(Exception::class) get() {
                if (com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.FirestoreUser.Companion.currentFirebaseUser == null) { throw com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.FirestoreUser.NullFirebaseUserException()
                }

                val db = FirebaseFirestore.getInstance()
                val userCollection = db.collection(com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.FirestoreUser.Companion.COLLECTION_USERS)
                return userCollection.document(com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.FirestoreUser.Companion.currentFirebaseUser!!.uid)
            }

        @Throws(Exception::class)
        fun buildUserDocument(): DocumentReference {
            val currentUserDoc =
                com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.FirestoreUser.Companion.userDocument

            val user = com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.FirestoreUser.Companion.currentFirebaseUser
                ?:
                throw com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.FirestoreUser.NullFirebaseUserException(
                    "Aborting creation of new User in database."
                )

            currentUserDoc.set(HashMap<String, Any>(), SetOptions.merge())
                .addOnSuccessListener {
                    Log.d("Firestore",
                        "User document ${ user.uid } successfully FOUND!")
                }
                .addOnFailureListener { e: Exception ->
                    Log.d("Firestore",
                        "User document ${ user.uid } could NOT be GENERATED / LOCATED!")
                    e.printStackTrace()
                }
                .addOnCompleteListener {
                    Log.d("Firestore",
                        "User document ${ user.uid } process complete.")
                    currentUserDoc.get().addOnSuccessListener { docSnap: DocumentSnapshot ->
                        if (docSnap.exists()) {
                            try { FirestoreAccount.init() }
                            catch (e: Exception) { e.printStackTrace() }
                            Log.d("Firestore",
                                "User document ${ user.uid } successfully INITIALIZED!")
                        }
                    }
                }

            return currentUserDoc
        }

        @TestOnly
        fun getCurrentFirebaseUserDisplayNameInitials(displayName: String?): String {
            val displayInitials = StringBuilder()
            if (displayName != null) {
                val names =
                    displayName.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

                for (name in names) {
                    val trimmedName = name.trim { it <= ' ' }
                    if (trimmedName.isNotEmpty()) {
                        val initial = trimmedName[0]
                        displayInitials.append(initial)
                        if (displayInitials.length >= 2) { break }
                    }
                }
            }
            return displayInitials.toString()
        }

    }

    class NullFirebaseUserException : Exception {
        constructor() : super(com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.FirestoreUser.NullFirebaseUserException.Companion.ERROR_MESSAGE)

        constructor(extendedMessage: String) : super("${com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.FirestoreUser.NullFirebaseUserException.Companion.ERROR_MESSAGE} $extendedMessage")

        companion object {
            private const val ERROR_MESSAGE = "FirebaseUser is null."
        }
    }
}